### PR TITLE
Reverse polarity of the off-policy measurement

### PR DESCRIPTION
**Pull request by tdene**
# What does this PR do ?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/inference/engines/dynamic_engine.py [hunk 1]
```diff
@@ -262,6 +262,7 @@ def reset(self) -> None:
         self.requests: Dict[int, RequestEntry] = {}
         self.waiting_request_ids = deque()
         self.failed_request_ids = []
+        self._generation_epoch: Optional[int] = None
```
Comment (yobibyte) [inline megatron/core/inference/engines/dynamic_engine.py:278]: In what situation are we okay with this being None?

Comment (tdene) [reply megatron/core/inference/engines/dynamic_engine.py:278]: When we're not doing RL.

File: megatron/core/inference/engines/dynamic_engine.py [hunk 2]
```diff
@@ -1132,6 +1133,34 @@ def post_process_requests(
...
+                                    torch.full(
+                                        (total - len(tensor),),
+                                        it,
+                                        dtype=tensor.dtype,
+                                        device=tensor.device,
+                                    ),
+                                ),
+                                dim=0,
+                            ),
+                        )
```
Comment (santhnm2) [inline megatron/core/inference/engines/dynamic_engine.py]: Can you move this to a helper function (e..g., `_set_generation_epoch`)?

File: megatron/core/inference/data_parallel_inference_coordinator.py [hunk 3]
```diff
@@ -447,7 +447,9 @@ def start(self):
                     self.state = self.CoordinatorState.STOPPING
 
                 # Broadcast the control signal if we're in a good state.
-                broadcast_payload = msgpack.packb([header.value], use_bin_type=True)
+                # Forward the full deserialized payload so that data-bearing
+                # signals (e.g. SET_GENERATION_EPOCH) retain their arguments.
+                broadcast_payload = msgpack.packb(deserialized_payload, use_bin_type=True)
```
Comment (santhnm2) [inline megatron/core/inference/data_parallel_inference_coordinator.py:450]: Does this add overhead?

Comment (tdene) [reply megatron/core/inference/data_parallel_inference_coordinator.py:450]: This was my main concern. I timed it. It does not add any measurable overhead, because it is still a single packet that's being sent over the network (just a slightly larger packet).

File: megatron/core/inference/engines/dynamic_engine.py (hunk 2 above)
Comment (tdene) [reply megatron/core/inference/engines/dynamic_engine.py]: Addressed! I rewrote the logic itself so that it's not awful code. Do you think this new version is readable enough that we can get rid of the `_set_generation_epoch` helper and just in-line this?

File: megatron/rl/rl_utils.py [hunk 4]
```diff
@@ -1895,15 +1862,14 @@ def megatron_rl_inference_mode(
             toggle_cuda_graphs(lang_module, cuda_graph_impl)
 
         inference_interface = get_inference_interface(args, loop, model)
+        inference_interface.set_generation_epoch(get_args().curr_iteration)
```
Comment (lmcafee-nvidia) [inline megatron/rl/rl_utils.py:1867]: if I'm understanding this correctly, the high level goal is update each token's 'staleness tracking' value only once on its first iteration, rather than updating/incrementing it for each iteration that the token is live, is that correct? are there any other goals here, such as knowing the training iteration (I don't know if this was tracked before)?

File: megatron/rl/rl_utils.py [hunk 5]
```diff
@@ -795,13 +794,13 @@ def compute_group_stats(
             roll_turn_lens = [len(t) for t in rollout.trajectory]
             group_turn_lengths.extend(roll_turn_lens)
             group_traj_lengths.append(sum(roll_turn_lens))
-            group_policy_staleness.extend(s for turn in rollout.policy_staleness for s in turn)
-            group_kv_staleness.extend(s for turn in rollout.kv_cache_staleness for s in turn)
-            group_completed_at_steps.extend(rollout.completed_at_step)
+            group_policy_epoch.append(min(s for turn in rollout.policy_epoch for s in turn))
```
Comment (lmcafee-nvidia) [inline megatron/rl/rl_utils.py]: is it possible for these lists to ever be empty? if so, the min/max ops on these lines need to be handled properly

File: megatron/core/inference/engines/dynamic_engine.py [hunk 6]
```diff
@@ -1160,6 +1165,25 @@ def post_process_requests(
 
         return active_request_ids, finished_request_records
 
+    @staticmethod
+    def _stamp_generation_epoch(request: DynamicInferenceRequest, epoch: int):
+        """Extend policy_epoch and kv_cache_epoch tensors to cover all current tokens."""
+        total = len(request.prompt_tokens) + len(request.generated_tokens)
+
+        if (t := request.policy_epoch) is not None:
+            request.policy_epoch = torch.cat(
```
Comment (lmcafee-nvidia) [inline megatron/core/inference/engines/dynamic_engine.py]: may be a small performance optimization to skip calling `torch.cat` if `total - len(t) == 0`. i.e., guard with a `if total - len(t) > 0:`

File: megatron/rl/rl_utils.py (hunk 4 above)
Comment (lmcafee-nvidia) [inline megatron/rl/rl_utils.py:1867]: can there be a race condition at all between the epoch signal and the resume signal on the next line. For example, can the resume signal arrive first, and tokens continue generating with the old epoch value, and then the updated epoch value arrives, in which case some tokens are stamped with the wrong epoch?

File: megatron/core/inference/engines/dynamic_engine.py (hunk 6 above)
Comment (lmcafee-nvidia) [inline megatron/core/inference/engines/dynamic_engine.py]: just an observation, each `torch.cat()` is an O(N) copy operation where N==generation length, and this `torch.cat()` is called O(N) times since it's called during `_post_process_requests()` for each generated token. so while `torch.cat()` is relatively cheap, we are doing O(N^2) operations for tracking staleness. are there any scalability implications to worry about?

File: megatron/rl/rl_utils.py (hunk 4 above)
Comment (tdene) [reply megatron/rl/rl_utils.py:1867]: > if I'm understanding this correctly, the high level goal is update each token's 'staleness tracking' value only once on its first iteration, rather than updating/incrementing it for each iteration that the token is live, is that correct?

Correct. When the token is generated, it gets stamped with a label. That label never changes.

> are there any other goals here, such as knowing the training iteration (I don't know if this was tracked before)?

Right now, in GRPO, we are stamping the token with the training iteration.    
We have been talking about other algorithms that will require other kinds of stamps... but this PR is future-facing and would let us implement such "other algorithms", because we're just stamping the token with a value.

Comment (tdene) [reply megatron/rl/rl_utils.py:1867]: There cannot be, specifically because both the epoch signal and resume signal are coded as "signal"s. That makes them enter a specific "signal" queue in the engine, which is guaranteed to arrive and be processed in order.

Comment (lmcafee-nvidia) [reply megatron/rl/rl_utils.py:1867]: got it, thanks

Comment (lmcafee-nvidia) [reply megatron/rl/rl_utils.py:1867]: makes sense!

File: megatron/rl/rl_utils.py (hunk 5 above)
Comment (tdene) [reply megatron/rl/rl_utils.py]: Handled with an assert.

File: megatron/core/inference/engines/dynamic_engine.py (hunk 6 above)
Comment (tdene) [reply megatron/core/inference/engines/dynamic_engine.py]: See the thread below about performance optimizations.

Comment (tdene) [reply megatron/core/inference/engines/dynamic_engine.py]: I will actually do better than fix this.

A different reviewer's feedback has led me to redesign the system, so that instead of storing massive tensors of repeated numbers, we just compress them from `[3, 3, 3, 4, 4, 4, 5, 5, 6, 6, 6, 6]` to `[ (3,0), (4, 3), (5, 6), (6, 8) ]`.

Comment (lmcafee-nvidia) [reply megatron/core/inference/engines/dynamic_engine.py]: great approach!

File: megatron/core/inference/engines/dynamic_engine.py [hunk 7]
```diff
@@ -828,6 +829,12 @@ def _add_request(
             )
             request.add_event_add_engine()  # Record when request enters engine
 
+            # Stamp new request with the current generation epoch.
+            if self._generation_epoch is not None:
+                epoch = self._generation_epoch
+                request.policy_epoch = [(0, epoch)]
```
Comment (yobibyte) [inline megatron/core/inference/engines/dynamic_engine.py:835]: Can generation_epoch be 0?

Comment (tdene) [reply megatron/core/inference/engines/dynamic_engine.py:835]: Yes, it can be. It could be any value inference's user (i.e. RL) specifies. Does that cause any issues here, I don't see.

Comment (yobibyte) [reply megatron/core/inference/engines/dynamic_engine.py:835]: NVM, I misread the meaning of the tuple. I first thought we will have two zeros there.
